zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorTest.java

```diff
@@ -481,6 +481,77 @@ public void shouldSetSourcePointerForFollowUpRecords() {
+    // given
+    final var defaultRecordProcessor = streamPlatform.getDefaultMockedRecordProcessor();
+
+    when(defaultRecordProcessor.process(any(), any()))
+        .thenAnswer(
+            (invocation) -> {
+              final var processingResultBuilder =
+                  (BufferedProcessingResultBuilder) invocation.getArgument(1);
+              processingResultBuilder.appendRecordReturnEither(
+                  3,
+                  Records.processInstance(3),
+                  new RecordMetadata()
+                      .recordType(RecordType.EVENT)
+                      .intent(ACTIVATE_ELEMENT)
+                      .rejectionType(RejectionType.NULL_VAL)
+                      .rejectionReason(""));
+              processingResultBuilder.appendRecordReturnEither(
+                  4,
+                  Records.processInstance(4),
+                  new RecordMetadata()
+                      .recordType(RecordType.COMMAND)
+                      .intent(COMPLETE_ELEMENT)
+                      .rejectionType(RejectionType.NULL_VAL)
+                      .rejectionReason(""));
+              processingResultBuilder.appendRecordReturnEither(
+                  5,
+                  Records.processInstance(5),
+                  new RecordMetadata()
+                      .recordType(RecordType.EVENT)
+                      .intent(ACTIVATE_ELEMENT)
+                      .rejectionType(RejectionType.NULL_VAL)
+                      .rejectionReason(""));
+              return processingResultBuilder.build();
+            })
+        .thenAnswer(
+            (invocation) -> {
+              final var processingResultBuilder =
+                  (BufferedProcessingResultBuilder) invocation.getArgument(1);
+              processingResultBuilder.appendRecordReturnEither(
+                  6,
+                  Records.processInstance(6),
+                  new RecordMetadata()
+                      .recordType(RecordType.EVENT)
+                      .intent(ACTIVATE_ELEMENT)
+                      .rejectionType(RejectionType.NULL_VAL)
+                      .rejectionReason(""));
+              return processingResultBuilder.build();
+            });
+
+    streamPlatform.startStreamProcessor();
+
+    // when
+    streamPlatform.writeBatch(
+        RecordToWrite.command().processInstance(COMPLETE_ELEMENT, Records.processInstance(2)));
+
+    // then
+    verify(defaultRecordProcessor, TIMEOUT.times(2)).process(any(), any());
+
+    final var logStreamReader = streamPlatform.getLogStream().newLogStreamReader();
+    logStreamReader.seekToFirstEvent();
+    final var firstRecord = logStreamReader.next();
+    assertThat(firstRecord.getSourceEventPosition()).isEqualTo(-1);
+
+    while (logStreamReader.hasNext()) {
+      final var record = logStreamReader.next();
+      assertThat(record.getSourceEventPosition()).isEqualTo(record.getKey() < 6 ? 1 : 3);
+    }
```

Comment:
:wrench: That's a bit too smart for my liking, I'd rather see the expected source event positions spelled out and checked for each record.

:wrench: This doesn't assert that we see the expected number of records, so the test could pass if there are records missing. This is solved too if we just read the 5 expected records and then assert for each of them that they have the correct source event position.

Reply:
I unrolled the loop then.